Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,286 @@
+
+open scoped Jsx in
+/-- Add the variable `v` with the type `tp` to the substance program. -/
+def addPenroseVar (tp : String) (v : PenroseVar) :
+    DiagramBuilderM Unit := do
+  let h := <InteractiveCode fmt={← Widget.ppExprTagged v.e} />
+  addEmbed (toString v) tp h
+
+/-- Add constructor `tp v := nm (vs)` to the substance program. -/
+def addConstructor (tp : String) (v : PenroseVar) (nm : String) (vs : List PenroseVar) :
+    DiagramBuilderM Unit := do
+  let vs' := ", ".intercalate (vs.map (fun v => toString v))
+  addInstruction s!"{tp} {v} := {nm} ({vs'})"
+
+open scoped Jsx in
+/-- Construct a string diagram from a Penrose `sub`stance program and expressions `embeds` to
+display as labels in the diagram. -/
+def mkStringDiagram (nodes : List (List Node)) (strands : List (List Strand)) :
+    DiagramBuilderM PUnit := do
+  /- Add 2-morphisms. -/
+  for x in nodes.join do
+    match x with
+    | .atom _ => do addPenroseVar "Atom" x.toPenroseVar
+    | .id _ => do addPenroseVar "Id" x.toPenroseVar
+  /- Add constraints. -/
+  for l in nodes do
+    for (x₁, x₂) in pairs l do
+      addInstruction s!"Left({x₁.toPenroseVar}, {x₂.toPenroseVar})"
+  /- Add constraints. -/
+  for (l₁, l₂) in pairs nodes do
+    if let .some x₁ := l₁.head? then
+      if let .some x₂ := l₂.head? then
+        addInstruction s!"Above({x₁.toPenroseVar}, {x₂.toPenroseVar})"
+  /- Add 1-morphisms as strings. -/
+  for l in strands do
+    for s in l do
+      addConstructor "Mor1" s.toPenroseVar
+        "MakeString" [s.startPoint.toPenroseVar, s.endPoint.toPenroseVar]
+
+/-- Penrose dsl file for string diagrams. -/
+def dsl :=
+  include_str ".."/".."/".."/"widget"/"src"/"penrose"/"monoidal.dsl"
+
+/-- Penrose sty file for string diagrams. -/
+def sty :=
+  include_str ".."/".."/".."/"widget"/"src"/"penrose"/"monoidal.sty"
+
+open scoped Jsx in
+/-- Construct a string diagram from the expression of a 2-morphism. -/
+def fromExpr (e : Expr) : MonoidalM Html := do
+  let e' := (← eval e).expr
+  DiagramBuilderM.run do
+    mkStringDiagram (← e'.nodes) (← e'.strands)
+    match ← DiagramBuilderM.buildDiagram dsl sty with
+    | some html => return html
+    | none => return <span>No String Diagram.</span>
+
+open scoped Jsx in
+/-- The RPC method for displaying string diagrams. -/
+@[server_rpc_method]
+def rpc (props : PanelWidgetProps) : RequestM (RequestTask Html) :=
+  RequestM.asTask do
+    let inner : Option (Html × Html) ← (do
+      if props.goals.isEmpty then
+        return none
+      let some g := props.goals[0]? | unreachable!
+
+      g.ctx.val.runMetaM {} do
+        let md ← g.mvarId.getDecl
+        let lctx := md.lctx
+        Meta.withLCtx lctx md.localInstances do
+          let e : Expr ← g.mvarId.getType
+          let some (_, lhs, rhs) := e.eq? | return none
+          let some ctx ← mkContext lhs | return none
+          MonoidalM.run ctx do
+            return some ⟨← fromExpr lhs, ← fromExpr rhs⟩)
```

Comment:
Optional: Looking at the examples I had the idea that you could modify `fromExpr` with an optional argument so that in the case of equality, you get the text `no string diagram for RHS`/`.. for LHS`.

I was for a split-second confused why it displays one diagram plus "No string diagram" in one of the examples.